site/content/blog/2021-04-23-optimizing-go-compiler-3.md

```diff
@@ -0,0 +1,234 @@
+---
+title: "Fixing a bug in the Go compiler as a newbie: a deep dive (III)"
+heading: "Fixing a bug in the Go compiler as a newbie: a deep dive (III)"
+description: ""
+summary: "Finally, the last part of the series! We have learned how the compiler uses Static Single Assignment and the rewrite rules to write and optimize intermediate code, and we have dissected each and every line of the SSA generated by the compiler in the buggy function. Now it is time to use all that knowledge to apply a fix, which will come in the form of the most simple rewrite rule."
+slug: optimizing-go-compiler-3
+date: 2021-04-23T12:02:00-05:00
+categories:
+    - "go"
+author: Alejandro García Montoro
+github: agarciamontoro
+community: alejandro.garcia
+---
+
+This is the third and last part of a series of posts on the compiler of Go. Make sure you have read [the first](/blog/optimizing-go-compiler-1) and [the second](/blog/optimizing-go-compiler-2) parts before going on.
+
+In the first part of the series, we discussed what the issue was, and in the second part we actually understood what was happening. Let's use all that knowledge to finally find a fix for the issue.
+
+# The fix
+
+We know now that our goal is to make sure that the generated code uses a single `MOVQstore`, instead of the smaller `MOVBstore`, `MOVWstore` and `MOVLstore` instructions. We have two ways of getting there:
+
+1. Trying to understand how the rules in the [generic.rules](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/gen/generic.rules) and [AMD64.rules](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/gen/AMD64.rules) files are applied during the compilation, and trying to fix the conflict that generates the weird pattern by adjusting the existing rules: maybe reordering them, maybe making them more specific or more generic, or maybe something else.
+2. Adding a new rule that explicitly matches that pattern and changes it with what we need.
+
+My first instinct was to go with 1: it is a more elegant approach, and would let us actually fix the issue, instead of patching it. It would also avoid the overhead of adding a new rule, which should be always checked and only applied when this specific, weird pattern happens.
```

Comment:
```suggestion
My first instinct was to go with 1: It's a more elegant approach, and would let us actually fix the issue, instead of patching it. It would also avoid the overhead of adding a new rule, which should be always checked and only applied when this specific, weird pattern happens.
```